### ISSUE TITLE
Code Coverage Incorrect for Parameter Default When Using New Profiler

### ISSUE THREAD

**Issue by wethreetrees** [pester/Pester#2119]
## General summary of the issue

When using the new code coverage profiler (`$PesterPreference.CodeCoverage.UseBreakpoints = $false`), Pester is reporting missing commands for parameter default value expressions.

ex:
```powershell
[Int]$Temperature = (Get-Random -Maximum 100 -Minimum 0)
```

This is working, as expected, when using breakpoints (`$PesterPreference.CodeCoverage.UseBreakpoints = $true`)

## Describe your environment

Pester version     : 5.3.1 C:\Client\PowerShell\Modules\Pester\5.3.1\Pester.psm1
PowerShell version : 7.2.1
OS version         : Microsoft Windows NT 10.0.19042.0

## Steps to reproduce

1. Download and extract the [PesterCCProfilerIssue.zip](https://github.com/pester/Pester/files/7747675/PesterCCProfilerIssue.zip)
2. From the target directory, run:
```powershell
.\Invoke-Tests.ps1
```

## Expected Behavior

Should return no missing commands with 100% code coverage.

## Current Behavior

Returns:
```powershell
Covered 75% / 75%. 4 analyzed Commands in 1 File.
 Missed command:
 
File            Class Function        Line Command
----            ----- --------        ---- -------
TestModule.psm1       Get-Temperature    6 Get-Random -Maximum 100 -Minimum 0
```

## Possible Solution? (optional)

😕

**Comment by fflaten** [comment]
Thanks for the report. I can confirm the issue.

The following lines might indicate it's a known limitation because we couldn't register a hit for parameter values either way.
https://github.com/pester/Pester/blob/13d65958ae6dc33bb18824b6acb48509f11b6784/src/functions/Coverage.ps1#L1213-L1218

CC debug of `Get-Temperature` (notice that the param block isn't hit).
```
PS /workspaces/Pester/Samples/PesterCCProfilerIssue> 
DBG: 9+     end  >>>> {
EXP: /workspaces/Pester/Samples/PesterCCProfilerIssue/TestModule.psm1:9:9:{
DBG: 10+         if ( >>>> $Temperature -gt 50) {
EXP: /workspaces/Pester/Samples/PesterCCProfilerIssue/TestModule.psm1:10:13:$Temperature -gt 50
DBG: 13+         return  >>>> 'Thats Cold'
EXP: /workspaces/Pester/Samples/PesterCCProfilerIssue/TestModule.psm1:13:16:'Thats Cold'
DBG: 14+      >>>> }
EXP: /workspaces/Pester/Samples/PesterCCProfilerIssue/TestModule.psm1:14:5:}
```

However if I use `Set-PSDebug -Trace 1` (the same events the profiler hooks into) it shows line 6 being called when default parameter value is used:
```
> Get-Temperature -Temperature 1
DEBUG:    1+  >>>> Get-Temperature -Temperature 1
DEBUG:    9+     end  >>>> {
DEBUG:   10+         if ( >>>> $Temperature -gt 50) {
DEBUG:   13+         return  >>>> 'Thats Cold'
Thats Cold
DEBUG:   14+      >>>> }

> Get-Temperature
DEBUG:    1+  >>>> Get-Temperature
** DEBUG:    6+         [Int]$Temperature =  >>>> (Get-Random -Maximum 100 -Minimum 0) **
DEBUG:    9+     end  >>>> {**
DEBUG:   10+         if ( >>>> $Temperature -gt 50) {
DEBUG:   13+         return  >>>> 'Thats Cold'
Thats Cold
DEBUG:   14+      >>>> }
```

@nohwnd, do you remember any of this? If powershell traces line 6, should we get it too? Are we excluding it for some reason?

**Comment by fflaten** [comment]
Oh, I see it. The message show the parameter value being calculated, but the extent we're able to get at that point is just the reference/call to the function.

![image](https://user-images.githubusercontent.com/3436158/146922018-d92fdb65-3e1c-4626-934c-4197584d27fa.png)

**Comment by nohwnd** [comment]
Yeah I think this was one of the limitations, there is a test that compares old cc with new, and comments on what we can do better than the previous CC and what we can do worse. This is probably one of the (few) things we do worse.

https://github.com/pester/Pester/blob/main/tst/CoverageTestFile.ps1
https://github.com/pester/Pester/blob/main/tst/Pester.RSpec.Coverage.ts.ps1#L30

**Comment by nohwnd** [comment]
Closing this as a known limitation. As fflaten and I worked out above, when a parameter default is evaluated the only extent the profiler can get at that point is the call/reference to the function, not the default-value expression itself, so we can't register a hit on that line. This is one of the few cases where the profiler-based code coverage does worse than the old breakpoint-based one, and it's already captured in our old-vs-new CC comparison (tst/CoverageTestFile.ps1 and tst/Pester.RSpec.Coverage.ts.ps1). I don't see a way to fix it without PowerShell giving us a better extent for that event, so I'll close it as a documented limitation. Thanks for the detailed repro!
